libp2p/errors.nim

```diff
@@ -37,3 +37,14 @@ macro checkFutures*[T](futs: seq[Future[T]], exclude: untyped = []): untyped =
             # We still don't abort but warn
             warn "Something went wrong in a future",
                 error=exc.name, file=pos.filename, line=pos.line
+
+proc allFuturesThrowing*[T](args: varargs[Future[T]]): Future[void] =
```

Comment:
so raising on first error is a bit arbitrary - it's a special case that happens to be interesting for us - conflating "check list of futures for failures" and "wait for list of futures" could perhaps be composed better: `if anyIt(await allFinished, it.failed):` or `filterIt` - seeing that, one could write a few "helpers" that deal with list-of-finished futures?

Reply:
I agree that throwing on first is arbitrary and we should start adding more of this helpers. Here is a link for ideas/inspiration - http://bluebirdjs.com/docs/api-reference.html, not endorsing this, just posting for reference. 

I disagree however that `if anyIt(await allFinished, it.failed):` will be a good solution, first it's error prone, second it becomes tedious and repetitive to write them by hand every time. The helpers, if well defined are convenient and prevent common sources of bugs in the long run.

Reply:
I on purpose made it in a "fail fast fail often" fashion cos this should not be used in production code (maybe we should put some big warning in release?)

Reply:
> tedious and repetitive

I meant it merely as an example to show that the problem is composable - of course, if it deserves it, there should be special helpers - that said, the `any`, `first/one` etc constructs are classics that are applicable here as well and we'd do well to tie into them when designing the api because they've stood the test of time in terms of applicability and generality

Reply:
Cool, then we agree :) 

In my experience, the helpers are handy and do add value, so maybe we should experiment with them in libp2p and then propose in chronos, or even ship as a separate library, dunno, but seems worth the effort.